miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,458 @@
+    }
+
+    /// Returns a forest with only the smallest tree present.
+    ///
+    /// If forest cannot be empty, use `smallest_tree` for performance.
+    pub fn smallest_tree(self) -> Self {
+        if self.is_empty() {
+            return Self::empty();
+        }
+        self.smallest_tree_unchecked()
+    }
+
+    /// Keeps only trees larger than the reference tree.
+    ///
+    /// For example, if we start with the bit pattern `0b0101_0110`, and keep only the trees larger
+    /// than tree index 1, that targets this bit:
+    /// ```text
+    /// Forest(0b0101_0110).trees_larger_than(1)
+    ///                        ^
+    /// Becomes:      0b0101_0100
+    ///                        ^
+    /// ```
+    /// And keeps only trees *after* that bit, meaning that the tree at `tree_idx` is also removed,
+    /// resulting in `0b0101_0100`.
+    ///
+    /// ```
+    /// # use miden_crypto::merkle::Forest;
+    /// let range = Forest::new(0b0101_0110);
+    /// assert_eq!(range.trees_larger_than(1), Forest::new(0b0101_0100));
+    /// ```
+    pub fn trees_larger_than(self, tree_idx: u32) -> Self {
+        self & high_bitmask(tree_idx + 1)
+    }
+
+    /// Creates a new forest with all possible trees smaller than the smallest tree in this
+    /// forest.
+    ///
+    /// This forest must have exactly one tree.
+    ///
+    /// # Panics
+    /// With debug assertions enabled, this function panics if this forest does not have
+    /// exactly one tree.
+    ///
+    /// For a non-panicking version of this function, see [`Forest::all_smaller_trees()`].
+    pub fn all_smaller_trees_unchecked(self) -> Self {
+        debug_assert_eq!(self.num_trees(), 1);
+        Self::new(self.0 - 1)
+    }
+
+    /// Creates a new forest with all possible trees smaller than the smallest tree in this
+    /// forest, or returns `None` if this forest has more or less than one tree.
+    ///
+    /// If the forest cannot have more or less than one tree, use
+    /// [`Forest::all_smaller_trees_unchecked()`] for performance.
+    pub fn all_smaller_trees(self) -> Option<Forest> {
+        if self.num_trees() != 1 {
+            return None;
+        }
+        Some(self.all_smaller_trees_unchecked())
+    }
+
+    /// Returns a forest with exactly one tree, one size (depth) larger than the current one.
+    pub fn next_larger_tree(self) -> Self {
+        debug_assert_eq!(self.num_trees(), 1);
+        Forest(self.0 << 1)
+    }
+
+    /// Returns true if the forest contains a single-node tree.
+    pub fn has_single_leaf_tree(self) -> bool {
+        self.0 & 1 != 0
+    }
+
+    /// Add a single-node tree if not already present in the forest.
+    pub fn single_leaf_tree_added(self) -> Self {
+        Self::new(self.0 | 1)
+    }
+
+    /// Remove the single-node tree if present in the forest.
+    pub fn single_leaf_tree_removed(self) -> Self {
+        Self::new(self.0 & (usize::MAX - 1))
+    }
```

Comment:
nit: should we use present tense in naming these methods? Maybe something like `add_single_leaf_tree()` and `remove_single_leaf_tree()`? Or maybe even `with_single_leaf_tree()` and `without_single_leaf_tree()`.

Reply:
I like with/without, since it returns a new value instead of mutating the existing one.